core/packages/contracts/contracts/NativeTokens.sol

```diff
@@ -0,0 +1,166 @@
+
+    /// @dev Initializes the NativeTokens contract with a vault and channels.
+    /// @param _vault The vault to use to `lock`/`unlock` tokens.
+    /// @param _outboundChannel The channel used to queue lock and create messages.
+    /// @param _allowedOrigin The origin allowed to call handle.
+    constructor(ERC20Vault _vault, OutboundChannel _outboundChannel, bytes32 _allowedOrigin) {
+        vault = _vault;
+        outboundChannel = _outboundChannel;
+        allowedOrigin = _allowedOrigin;
+    }
+
+    /// @dev Locks an amount of ERC20 Tokens in the vault and enqueues a mint message. Requires the allowance to be set on the ERC20 token where the spender is the vault.
+    /// @param token The token to lock.
+    /// @param recipient The recipient on the substrate side.
+    /// @param amount The amount to lock.
+    function lock(address token, bytes32 recipient, uint128 amount) public {
+        if (amount == 0) {
+            revert ZeroAmount();
+        }
+        if (token == address(0)) {
+            revert ZeroAddressToken();
+        }
+        if (recipient == bytes32(0)) {
+            revert ZeroAddressRecipient();
+        }
+
+        // TODO: Implement a max locked amount.
+        vault.deposit(msg.sender, token, amount);
+
+        // TODO: Encode a call
+        bytes memory call;
+        // TODO: Get weight
+        uint64 weight = 1_000_000;
+
+        emit Locked(msg.sender, recipient, token, amount);
+        outboundChannel.submit(msg.sender, call, weight);
+    }
+
+    /// @dev Enqueues a create native token message to substrate.
+    /// @param token The ERC20 token address.
+    function create(address token) public {
+        if (token == address(0)) {
+            revert ZeroAddressToken();
+        }
+
+        IERC20Metadata metadata = IERC20Metadata(token);
+        // TODO: Use metadata to encode the call.
+        string memory name = metadata.name();
+        string memory symbol = metadata.symbol();
+        uint8 decimals = metadata.decimals();
+
+        // TODO: Encode a call
+        bytes memory call;
+        // TODO: Get weight
+        uint64 weight = 1_000_000;
+
+        emit Created(token, name, symbol, decimals);
+        outboundChannel.submit(msg.sender, call, weight);
+    }
+
+    /// @dev Processes messages from inbound channel.
+    /// @param origin The hashed substrate sovereign account.
+    /// @param message The message enqueued from substrate.
+    function handle(bytes32 origin, bytes calldata message) external onlyOwner {
+        if (origin != allowedOrigin) {
+            revert UnauthorizedOrigin();
+        }
+
+        Message memory decoded = abi.decode(message, (Message));
+        if (decoded.action == Action.Unlock) {
+            doUnlock(origin, abi.decode(decoded.payload, (UnlockPayload)));
+        } else {
+            revert UnsupportedMessageAction();
+        }
+    }
+
+    /// @dev Unlocks funds from the vault and sends it to recipient.
```

Comment:
We don't need docstrings for internal methods. Please remove. They just add extra noise.

Reply:
As I said before, lets not focus on docstrings at this stage.